jib-core/src/main/java/com/google/cloud/tools/jib/api/JibContainerBuilder.java

```diff
@@ -0,0 +1,259 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.api;
+// TODO: Move to com.google.cloud.tools.jib
+
+import com.google.cloud.tools.jib.configuration.LayerConfiguration;
+import com.google.cloud.tools.jib.configuration.Port;
+import com.google.cloud.tools.jib.image.ImageReference;
+import com.google.common.collect.ImmutableList;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * Builds a container with Jib.
+ *
+ * <p>Example usage:
+ *
+ * <pre>{@code
+ * Jib.from(baseImage)
+ *    .addFiles(sourceFiles, extractionPath)
+ *    .setEntrypoint(entrypoint)
+ *    .setProgramArguments(arguments)
+ *    .addEnvironmentVariable(key, value)
+ *    .addExposedPort(
+ *    .addLabel(key, value)
+ *    .containerize(...);
+ * }</pre>
+ */
+public class JibContainerBuilder {
+
+  private final ImageReference baseImageReference;
+
+  private List<LayerConfiguration> layerConfigurations = new ArrayList<>();
+  private Map<String, String> environment = new HashMap<>();
+  private List<Port> ports = new ArrayList<>();
+  private Map<String, String> labels = new HashMap<>();
+  @Nullable private ImmutableList<String> entrypoint;
+  @Nullable private ImmutableList<String> programArguments;
+
+  /** Instantiate with {@link Jib#from}. */
+  JibContainerBuilder(ImageReference baseImageReference) {
+    this.baseImageReference = baseImageReference;
+  }
+
+  /**
+   * Adds a new layer to the container with {@code files} as the source files and {@code
+   * pathInContainer} as the path to copy the source files to in the container file system.
+   *
+   * <p>Source files that are directories will be recursively copied. For example, if the source
+   * files are:
+   *
+   * <ul>
+   *   <li>{@code fileA}
+   *   <li>{@code fileB}
+   *   <li>{@code directory/}
+   * </ul>
+   *
+   * and the destination to copy to is {@code /path/in/container}, then the new layer will have the
+   * following entries for the container file system:
+   *
+   * <ul>
+   *   <li>{@code /path/in/container/fileA}
+   *   <li>{@code /path/in/container/fileB}
+   *   <li>{@code /path/in/container/directory/}
+   *   <li>{@code /path/in/container/directory/...} (all contents of {@code directory/})
+   * </ul>
+   *
+   * @param files the source files to copy to a new layer in the container
+   * @param pathInContainer the Unix-style path to copy the source files to in the container file
+   *     system
+   * @return this
+   */
+  public JibContainerBuilder addFiles(List<Path> files, String pathInContainer) {
```

Comment:
Hmm maybe this should just be an overload of `addLayer` instead of being named differently?

Reply:
That sounds good, condenses the method names.